### PR TITLE
Reuse response (if any) to avoid sending an additional request.

### PR DESCRIPTION
**Pull request by alejo-enix**
**Description:**
If a response with a calendar component is available return it instead of requesting the resource by making another request to the server.

**Features added:**
Avoids making an additional request to the CalDAV server by reusing the existent response retrieved previously.

**How it is implemented:**
Refactor an existed method by adding a parameter to it, the parameter is a _MultiStatusResponse_ object which was previously retrieved when a request was executed then the object is validated to check if an additional request is needed anyways for the CalDAV resource to be fetched successfully.

### CODE REVIEW

**Comment by ioggstream** [comment]
@alejo-enix thanks for this PR. Can you please provide:

- a description of this PR
- what feature it adds to the library
- how it is implemented
- tests so that users can learn how to use this functionality

Thanks again, R.

**Comment by ioggstream** [comment]
close and reopen for CI

**Comment by TheAntimist** [comment]
Looks like some tests are failiing, could you look into them?

```17:09:22.980 [main] INFO  com.github.caldav4j.CalDAVCollectionTest - Check if event is still on server
[ERROR] Tests run: 14, Failures: 1, Errors: 2, Skipped: 3, Time elapsed: 7.298 s <<< FAILURE! - in com.github.caldav4j.CalDAVCollectionTest
[ERROR] testGetWithoutCacheThenWithCache(com.github.caldav4j.CalDAVCollectionTest)  Time elapsed: 0.593 s  <<< FAILURE!
java.lang.AssertionError: expected null, but was:<com.github.caldav4j.exceptions.CalDAV4JException: Problem executing method>
	at com.github.caldav4j.CalDAVCollectionTest.testGetWithoutCacheThenWithCache(CalDAVCollectionTest.java:353)

[ERROR] testGetEventResources(com.github.caldav4j.CalDAVCollectionTest)  Time elapsed: 1.517 s  <<< ERROR!
com.github.caldav4j.exceptions.CalDAV4JException: Problem executing method
	at com.github.caldav4j.CalDAVCollectionTest.testGetEventResources(CalDAVCollectionTest.java:218)
Caused by: java.lang.NullPointerException
	at com.github.caldav4j.CalDAVCollectionTest.testGetEventResources(CalDAVCollectionTest.java:218)

[ERROR] queryCalendar(com.github.caldav4j.CalDAVCollectionTest)  Time elapsed: 0.56 s  <<< ERROR!
com.github.caldav4j.exceptions.CalDAV4JException: Problem executing method
	at com.github.caldav4j.CalDAVCollectionTest.queryCalendar(CalDAVCollectionTest.java:142)
Caused by: java.lang.NullPointerException
	at com.github.caldav4j.CalDAVCollectionTest.queryCalendar(CalDAVCollectionTest.java:142)```

File: src/main/java/com/github/caldav4j/CalDAVCollection.java [hunk 1]
```diff
@@ -605,9 +606,12 @@ protected CalDAVResource getCalDAVResource(HttpClient httpClient,
 			}
 		}
 
-		//either the etag was old, or it wasn't in the cache so let's get it
-		//from the server       
-		return getCalDAVResourceFromServer(httpClient, path);
+		// ETag was old or it wasn't in the cache so let's get it from the server if we don't have it.
+		if (response != null && CalendarDataProperty.getCalendarfromResponse(response) != null) {
+			return new CalDAVResource(response);
+		} else {
```
Comment (ioggstream) [inline src/main/java/com/github/caldav4j/CalDAVCollection.java]: I'd remove the `else`

**Comment by ioggstream** [comment]
Thanks for the tests. IIUC you are explicitly calling the new interface ( eg where you explicitly pass the response).

Is there a test where the caldav4j uses the cached response directly without the user explicit asking for that?

**Comment by ioggstream** [comment]
Let's move it to devel branch

**Pull request closed without merging**